src/stream/src/common/column_mapping.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+
+pub struct StateTableColumnMapping {
+    /// index: state table column index, value: upstream column index
+    upstream_columns: Vec<usize>,
+    /// key: upstream column index, value: state table column index
+    mapping: HashMap<usize, usize>,
+}
+
+impl StateTableColumnMapping {
+    /// Creates a new column mapping with the upstream columns included in state table.
+    pub fn new(upstream_columns: Vec<usize>) -> Self {
+        let mapping = upstream_columns
+            .iter()
+            .enumerate()
+            .map(|(i, col_idx)| (*col_idx, i))
+            .collect();
+        Self {
+            upstream_columns,
+            mapping,
+        }
+    }
+
+    /// Convert state table column index to upstream chunk column index.
+    pub fn state_table_to_upstream(&self, idx: usize) -> Option<usize> {
+        self.upstream_columns.get(idx).copied()
+    }
+
+    /// Convert upstream chunk column index to state table column index.
+    pub fn upstream_to_state_table(&self, idx: usize) -> Option<usize> {
+        self.mapping.get(&idx).copied()
+    }
+
+    /// Return the number of columns in the mapping.
+    pub fn size(&self) -> usize {
```

Comment:
Better to name it `len` here. 😄
```suggestion
    pub fn len(&self) -> usize {
```

Reply:
ok